File: ts/src/agent/mapPlanModels.ts
export interface MapPromptProfile {
  originalPrompt: string;
  title: string;
  theme: string;
  difficulty: string;
  targetRoomCount: number;
  requestedKeys: string[];
  requestedSetPieces: string[];
  lightingHint: string;
}

export interface MapPlan {
  profile: MapPromptProfile;
  topology: RoomNode[];
  connections: ConnectionPlan[];
  rooms: RoomPlan[];
  validation: ValidationReport;
}

export interface RoomNode {
  id: string;
  label: string;
  role: string;
  required: boolean;
  lock?: string;
  grantsKey?: string;
  notes: string;
}

export interface ConnectionPlan {
  from: string;
  to: string;
  kind: string;
  requiredKey?: string;
  notes: string;
}

export interface RoomPlan {
  id: string;
  shape: string;
  width: number;
  height: number;
  floorHeight: number;
  ceilingHeight: number;
  light: number;
  floorTexture: string;
  ceilingTexture: string;
  wallTexture: string;
  encounterHint: string;
  traversalHint: string;
  landmark: string;
}

export interface ValidationReport {
  passed: boolean;
  score: number;
  hardIssues: string[];
  softIssues: string[];
}

export function defaultValidationReport(): ValidationReport {
  return {
    passed: false,
    score: 0,
    hardIssues: [],
    softIssues: [],
  };
}

File: WADTool/src/agent/MapPlanModels.kt
package wadtool.agent

import kotlinx.serialization.Serializable

@Serializable
data class MapPromptProfile(
    val originalPrompt: String,
    val title: String,
    val theme: String,
    val difficulty: String,
    val targetRoomCount: Int,
    val requestedKeys: List<String>,
    val requestedSetPieces: List<String>,
    val lightingHint: String
)

@Serializable
data class MapPlan(
    val profile: MapPromptProfile,
    val topology: List<RoomNode>,
    val connections: List<ConnectionPlan>,
    val rooms: List<RoomPlan> = emptyList(),
    val validation: ValidationReport = ValidationReport()
)

@Serializable
data class RoomNode(
    val id: String,
    val label: String,
    val role: String,
    val required: Boolean,
    val lock: String? = null,
    val grantsKey: String? = null,
    val notes: String = ""
)

@Serializable
data class ConnectionPlan(
    val from: String,
    val to: String,
    val kind: String,
    val requiredKey: String? = null,
    val notes: String = ""
)

@Serializable
data class RoomPlan(
    val id: String,
    val shape: String,
    val width: Int,
    val height: Int,
    val floorHeight: Int,
    val ceilingHeight: Int,
    val light: Int,
    val floorTexture: String,
    val ceilingTexture: String,
    val wallTexture: String,
    val encounterHint: String,
    val traversalHint: String,
    val landmark: String
)

@Serializable
data class ValidationReport(
    val passed: Boolean = false,
    val score: Int = 0,
    val hardIssues: List<String> = emptyList(),
    val softIssues: List<String> = emptyList()
)

interface MapSubagent<I, O> {
    val name: String
    fun run(input: I): O
}

